### ISSUE TITLE
Improve and unify structure of TSPAlgorithms

### ISSUE THREAD

**Issue by HannesWell** [jgrapht/jgrapht#872]
```
 * JGraphT version: 1.3.2-Snaphshot
 * Java version (java -version)/platform:  all
```
### Issue

**Improve `HamiltonianCycleAlgorithm` implementations**

With the new `HamiltonianCycleAlgorithm` implementations came the class `HamiltonianCycleAlgorithmBase` that the new implementations use as common super class. I suggest that the previously existing algorithms extend that base-class too. 

To improve the separation of the actual algorithm and boilerplate code I suggest to implement `getTour()` in the base class and move the common code for all algorithms like input checks, handling of single vertex graphs and potential multiple passes of the actual algorithm there.
More important I suggest to add two abstract methods to `HamiltonianCycleAlgorithmBase`:

- `protected abstract void initialize(Graph<V, E> graph);`
- `protected abstract GraphPath<V, E> computeTour();`

The method `initialize(Graph<V, E>)` is used to initialize the algorithm for the given graph. For example in `TwoOptHeuristicTSP` that method would be used to compute the distances matrix like it is now done in `TwoOptHeuristicTSP.init()`.
The method `computeTour()` is then used to compute one tour (in one pass) of the actual algorithm.
The `getTour()` implementation in `HamiltonianCycleAlgorithmBase` would then be something like this:
```
public GraphPath<V, E> getTour(Graph<V, E> graph)
{
    checkGraph(graph);

    // Handle a graph with single vertex
    if (graph.vertexSet().size() == 1) {
        return getSingletonTour(graph);
    }
    initializeAlgorithm(graph);

    GraphPath<V, E> shortestPath = null;
    double shortestPathLength = Double.POSITIVE_INFINITY;
    for (int i = 0; i < runs; i++) {
        GraphPath<V, E> path = computeTour();
        if (path.getWeight() < shortestPathLength) {
            shortestPath = path;
            shortestPathLength = path.getWeight();
        }
    }
    return shortestPath;
}
```
Maybe a seed vertex in `computeTour()` is also useful for other classes than the `NearestNeighborHeuristicTSP`. If this is the case the base class could also pass a different seed vertex in each pass. If not this can be done in `NearestNeighborHeuristicTSP`.
If a `HamiltonianCycleAlgorithm` can create different tours for the same graph in each pass, like `RandomTourTSP` or `NearestNeighborHeuristicTSP` with a variating seed, that implementation can set `passes>1` when calling the super constructor and offer it as constructor parameter. Otherwise they set it fixed `passes = 1` in their constructor.

**Improve `HamiltonianCycleImprovementAlgorithm` implementations**

I suggest to introduce a `HamiltonianCycleImprovementAlgorithmBase` that extends `HamiltonianCycleAlgorithmBase` and is the equivalent for `HamiltonianCycleImprovementAlgorithm` implementations (all extend it).
I would also implement `improveTour()` in the base class and have an abstract method 

- `protected abstract GraphPath<V, E> computeTourImprovement(GraphPath<V, E> tour);`

with the same intend as `computeTour()` in `HamiltonianCycleImprovementAlgorithmBase`.
Like in the current `TwoOptHeuristicTSP` class this class should have an `initializer`-`HamiltonianCycleAlgorithm` object used to compute the tour that is then improved. The computeTour() method of that base class would look similar to this:

```
@Override
protected GraphPath<V, E> computeTour()
{
    GraphPath<V, E> tour;
    if (initializer instanceof HamiltonianCycleAlgorithmBase) {
        tour = baseInitializer.computeTour();
    } else {
        tour = initializer.getTour(graph);
    }
    // Multiple tries should not be necessary here? Otherwise there should be a distinction between initialization tries and improvement tries.
    return computeImprovedTour(tour);
}
```
Previously the initializer is also initialized in the `initialize()` method of the ImprovementAlgorithm.

The key-point is that on this way multiple passes of initial tour computation and subsequent improvement can be done with only one initialization of initializer and improvement algorithm and without risking inconsistencies because the graph changed in the meantime. It enables a tight interaction of initializer and improvement while maintaining a loose coupling between the algorithms.
These suggestions are inspired by the idea to improve the interaction of a `NearestNeighborHeuristicTSP` initializer with subsequent improvement by `TwoOptHeuristicTSP`.

I already started a little bit of work (mainly for the base classes and the two mentioned implementations) and wanted your opinion before I apply it to the other algorithms and finish the coding.

**Comment by Toptachamann** [comment]
1. The idea about utilizing the `HamiltonianCycleAlgorithmBase` seems reasonable as currently, we have code duplications in the `TwoApproxMetricTSP` because of this. 

2. Adding more methods to the base class seems to be unnecessary. If we add more implementations, they might not fit into the created abstraction and we will end up overriding those methods. For example, the `PalmerHamiltonianCycle` algorithm doesn't require the graph to be complete, so calling `checkGraph()` is not applicable in this case.  At the moment, what we need to do is to move the common methods and the graph checking to the base class to avoid code duplication as much as possible.  

3. Regarding the `HamiltonianCycleImprovementAlgorithmBase`, I know that the policy for base classes is that they're needed if we have multiple algorithms that extend them. At the moment, there's only one implementation (`TwoOptHeuristicTSP`), so I would avoid adding a base class as this might result in future difficulties. Either way, this won't change anything for the users.

**Comment by HannesWell** [comment]
Alright, then I'll only apply your first point and avoid code duplications as much as possible.
I see your point regarding my further suggestions and agree that it could make it difficult for future contributions.

One more thing:
What do you think about providing setters for the fields of the `TwoOptHeuristicTSP` and `NearestNeighborHeuristicTSP` to reduce the number of constructors? The number of constructors is quite high because of the field-combinations they cover. Actually all fields of both algorithms could be changed by a setter, between calls to `getTour()` without any problem. With that the number of constructors could reduced to one without any parameter (for `TwoOptHeuristicTSP` maybe the `minCostImprovement` could still be a field assigned in the constructor because it is more a property of the algorithm rather than a property of a run of the algorithm. But technically there is no reason to not modify it in a setter).

**Comment by d-michail** [comment]
I also agree with @Toptachamann!

**Comment by Toptachamann** [comment]
Resolved via #926
